### PR TITLE
bail out of logging early

### PR DESCRIPTION
**Pull request by splitice**
**Very** Low hanging fruit. 

Accounts for 2% of RX perf trace on Vultr Single Core 1GB VM.

### CODE REVIEW

**Comment by splitice** [comment]
Found while profiling for #110

File: src/logger.c [hunk 1]
```diff
@@ -40,59 +40,62 @@ static const char *logident = NULL;
...
 int umbilical = 0;
 
+static bool should_log(int level){
+	// Bail out early if there is nothing to do.
+
+	if(!logcontrol && (level > debug_level || logmode == LOGMODE_NULL)) {
+		return false;
+	}
+
+	return (level > debug_level);
```
Comment (gsliepen) [inline src/logger.c:50]: That should be `return level <= debug_level`.

File: src/logger.c [hunk 2]
```diff
@@ -125,6 +128,8 @@ void logger(int level, int priority, const char *format, ...) {
 	va_list ap;
 	char message[1024] = "";
 
+	if(!should_log(level)) return;
```
Comment (gsliepen) [inline src/logger.c:131]: Please run `make astyle` from the top directory.

**Review by gsliepen — CHANGES_REQUESTED** [review]
With this patch, nothing is logged anymore on my machines.

File: src/logger.c (hunk 1 above)
Comment (splitice) [reply src/logger.c:50]: how did I miss this :(

fixed.

File: src/logger.c (hunk 2 above)
Comment (splitice) [reply src/logger.c]: done

File: src/logger.c [hunk 3]
```diff
@@ -40,59 +40,62 @@ static const char *logident = NULL;
...
 int umbilical = 0;
 
+static bool should_log(int level) {
+	// Bail out early if there is nothing to do.
+
+	if(!logcontrol && (level > debug_level || logmode == LOGMODE_NULL)) {
+		return false;
+	}
+
+	return level <= debug_level;
```
Comment (gsliepen) [inline src/logger.c:50]: This turns out to be wrong too. If `logcontrol == true`, the value of `level` is not necessarily the same debug level requested over a control channel. So we should `return true` here, or just simplify this whole function to:

```
return (level <= debug_level && logmode != LOGMODE_NULL) || logcontrol;
```

File: src/logger.c [hunk 4]
```diff
@@ -40,59 +40,62 @@ static const char *logident = NULL;
...
-		return;
-	}
-
-	if(level <= debug_level) {
-		switch(logmode) {
-		case LOGMODE_STDERR:
-			fprintf(stderr, "%s\n", message);
-			fflush(stderr);
-			break;
+	switch(logmode) {
```
Comment (gsliepen) [inline src/logger.c:61]: There needs to be an `if (level <= debug_level)` around this switch and the first `if`-statement after it, in order to correctly handle the case where `logcontrol == true`.

**Review by gsliepen — CHANGES_REQUESTED** [review]

File: src/logger.c (hunk 3 above)
Comment (splitice) [reply src/logger.c]: Updated with the code you provided.

File: src/logger.c (hunk 4 above)
Comment (splitice) [reply src/logger.c:55]: FYI at no point in the code is `logcontrol` documented. Unless you were a seasoned developer I would expect it to be an unknown code quirk.

Comment (splitice) [reply src/logger.c]: And is there a reason that REQ_LOG doesnt just change the log level? Slight difference in behaviour but simpler and potentially more powerful.

Comment (gsliepen) [reply src/logger.c]: You're right that the function of `logcontrol` is not well documented. There are multiple log outputs possible at the same time, with different log levels. The normal log output (to `stdout`, a file or to the syslog) is controlled by the global variable `debug_level`. However, you can use the `tinc log <level>` command to connect to a running tinc daemon and get a stream of log messages with a different log level. Whenever there is at least one control connection requesting log information, `logcontrol` is `true`. That's when we have to do extra work for every call to `logger()` to find out which connections want which log message.

Comment (gsliepen) [reply src/logger.c]: There's also `REQ_SET_DEBUG` which allows you to change the global log level of a running tinc daemon.

**Comment by gsliepen** [comment]
Cherry-picked as a459e57de5fa4d32ee1162e49770fd8730e3707d, thanks!

**Pull request closed without merging**